packages/editor/src/optimizer/common.ts

```diff
@@ -0,0 +1,614 @@
+ *    or will continue.
+ *  - `"Done"`: optimizer finished normally
+ *  - `OptimizerError`: optimizer errored and cannot continue.
+ */
+export type StepSequenceState = "Pending" | "Done" | OptimizationError;
+
+/**
+ * Info about a step sequence.
+ *  - `LayoutStats`: stage names and steps
+ *  - `parent`: The step sequence and step from which this sequence originated,
+ *    or null if this is a root sequence
+ *  - `state`: State of the step sequence
+ *  - `variation`: Variation string on last resample
+ */
+export type StepSequenceInfo = {
+  layoutStats: LayoutStats;
+  parent: HistoryLoc | null;
+  state: StepSequenceState;
+  variation: string;
+};
+
+/**
+ * Map from `StepSequenceID`s to `StepSequenceInfo`s
+ */
+export type HistoryInfo = Map<StepSequenceID, StepSequenceInfo>;
+
+/** Type of a resolve for a promise waiting on a `MessageResult` */
+export type MessageResolve = (result: MessageResult) => void;
+
+/**
+ * A `RenderState`, but without rendered labels which cannot be sent across
+ * threads. Use `layoutStateToRenderState` to combine with an `svgCache` to
+ * generate a `RenderState` for rendering.
+ */
+export type LayoutState = {
+  canvas: Canvas;
+  shapes: Shape<number>[];
+  labelMeasurements: LabelMeasurements;
+  variation: string;
+};
+
+/** Minimal state needed for rendering */
+export type RenderState = {
+  canvas: Canvas;
+  shapes: Shape<number>[];
+  labelCache: LabelCache;
+  variation: string;
+};
+
+/** Info for a successful compile */
+export type CompileInfo = {
+  diagramId: DiagramID;
+  warnings: PenroseWarning[];
+};
+
+// Message types
+
+/** Requests sent from main thread to broker or broker to worker */
+export type MessageRequest = {
+  tag: "MessageRequest";
+  messageId: MessageID;
+  data: MessageRequestData;
+};
+
+/** Responses to `MessageRequest`s */
+export type MessageResponse = {
+  tag: "MessageResponse";
+  messageId: MessageID;
+  result: MessageResult;
+};
+
+/** Unprompted notifications between threads. No response can be sent. */
+export type Notification = {
+  tag: "Notification";
+  data: NotificationData;
+};
+
+/** Data sent in a `MessageRequest` */
+export type MessageRequestData =
+  | CompileRequestData
+  | PollRequestData
+  | ComputeLayoutRequestData
+  | ResampleRequestData;
+
+/** Result type contained in a `MessageResponse` */
+export type MessageResult =
+  | CompileResult
+  | PollResult
+  | ComputeLayoutResult
+  | ResampleResult;
+
+/** Data contained in a `Notification` */
+export type NotificationData =
+  | InitData
+  | LabelMeasurementData
+  | DiscardDiagramData;
+
+/** Ensure request data and result pairs have the same tag, to check validity */
+export enum MessageTags {
+  Compile = "Compile",
+  Poll = "Poll",
+  ComputeLayout = "ComputeLayout",
+  DiscardDiagram = "DiscardDiagram",
+  DiscardStepSequence = "DiscardStepSequence",
+  Init = "Init",
+  LabelMeasurements = "LabelMeasurements",
+  Resample = "Resample",
+}
```

Comment:
Using an `enum` seems like a good idea! Is there a reason you don't do it for other `tag`s too?

Reply:
I wanted to enure that `MessageRequestData`isubtypes have the same tag as their corresponding `MessageResult` subtype, and so the enum really was necessary to prevent a typo in the definition of the types. Mistyping a tag during instantiation will just get caught by the typechecker, so I thought it was less necessary. The only other tags in this file are for `MessageResponse`, `MessageRequest, and `Notification`, but I can add those to a `MessageTypeTag` enum, if you'd like?

Reply:
Up to you! Sounds like it's not actually necessary as you mentioned; I was just curious.